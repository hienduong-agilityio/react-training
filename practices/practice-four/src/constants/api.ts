export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ROOT_API = import.meta.env.VITE_API_ENDPOINT;

export const API_URL = {
  PROJECT: `${ROOT_API}/projects`
};
