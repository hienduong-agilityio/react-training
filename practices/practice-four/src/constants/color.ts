// Consider using enums instead of constants
export const COLORS_CLASS = {
  DEFAULT: 'color-default',
  PRIMARY: 'color-primary',
  SUCCESS: 'color-success',
  WARNING: 'color-warning',
  DANGER: 'color-danger '
};
