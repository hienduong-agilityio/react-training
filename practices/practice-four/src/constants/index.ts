// Index constants files
export * from './api';
export * from './color';
export * from './filerSearchOption';
export * from './route';
export * from './rowPerPage';
export * from './store';
export * from './validationMessages';
