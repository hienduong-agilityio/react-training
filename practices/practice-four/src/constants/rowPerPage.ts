export const ROWS_PER_PAGE_OPTIONS = [
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 30, label: '30' },
  { value: 40, label: '40' },
  { value: 50, label: '50' }
];
