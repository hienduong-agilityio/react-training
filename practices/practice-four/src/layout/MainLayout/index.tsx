// Libraries
import { Outlet } from 'react-router-dom';

// Layouts
import { Sidebar, Navbar, TabLayout } from '@/layout';

/**
 * MainLayout component
 *
 * @returns {JSX.Element} The main layout structure with navigation components and content area.
 */
export const MainLayout = (): JSX.Element => {
  return (
    <main className='flex'>
      <Sidebar />
      <div className='flex flex-col w-full'>
        <Navbar />
        <TabLayout />
        <Outlet />
      </div>
    </main>
  );
};
