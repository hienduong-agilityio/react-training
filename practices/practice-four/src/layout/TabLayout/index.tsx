// Libraries
import { generateBreadcrumbItems } from '@/helpers/breadcrumb';
import { useMemo } from 'react';
import { useLocation, useParams } from 'react-router-dom';

// Components
import { Breadcrumbs } from '@/components';

/**
 * TabLayout component
 *
 * @returns {JSX.Element} - The TabLayout element
 */
export const TabLayout = (): JSX.Element => {
  const location = useLocation();
  const { id: projectId } = useParams();

  const breadcrumbItems = useMemo(
    () => generateBreadcrumbItems(location.pathname, projectId),
    [location.pathname, projectId]
  );

  return (
    <div className='px-4 py-2'>
      <Breadcrumbs items={breadcrumbItems} />
    </div>
  );
};
