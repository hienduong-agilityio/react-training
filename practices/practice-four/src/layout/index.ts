// Index Layout files
export * from './MainLayout';
export * from './NavBar';
export * from './ProjectLayout';
export * from './SideBar';
export * from './TabLayout';
