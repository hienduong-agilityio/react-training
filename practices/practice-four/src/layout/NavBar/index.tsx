// Libraries
import { Link } from 'react-router-dom';

// Components
import { Avatar } from '@/components';

// Svg
import notifyIcon from '@public/images/notificationIcon.svg';
import helpIcon from '@public/images/helpIcon.svg';

// Constants
import { ROUTE } from '@/constants';

/**
 * Navbar components
 *
 * @returns {JSX.Element} The Navbar component.
 */
export const Navbar = (): JSX.Element => {
  return (
    <nav className='bg-light w-full border'>
      <div className='flex flex-wrap items-center justify-between p-4'>
        <Link to={ROUTE.PROJECT} className='text-2xl font-semibold text-gray-900'>
          Projects
        </Link>
        <div className='flex items-center gap-6'>
          <img src={notifyIcon} className='cursor-not-allowed' alt='Notification' />
          <img src={helpIcon} className='cursor-not-allowed' alt='Help' />
          <Avatar customClass='rounded-full cursor-not-allowed' name='HienDuong' />
        </div>
      </div>
    </nav>
  );
};
