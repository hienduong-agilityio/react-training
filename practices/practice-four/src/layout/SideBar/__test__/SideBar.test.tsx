// Libraries
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

// Components
import { Sidebar } from '@/layout';

// Constants
import { ROUTE } from '@/constants';

// Mock the SVG imports
jest.mock('@public/images/logoIcon.svg', () => 'mock-logo');
jest.mock('@public/images/projectsIcon.svg', () => 'mock-projects-icon');
jest.mock('@public/images/settingIcon.svg', () => 'mock-setting-icon');

/*
SideBar
  √ should match snapshot
  √ should render the SideBar children
*/
describe('SideBar', () => {
  const renderSideBar = () => {
    return render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
  };

  it('should match snapshot', () => {
    const { asFragment } = renderSideBar();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the SideBar children', () => {
    renderSideBar();

    const logo = screen.getByAltText('Logo');
    expect(logo).toBeInTheDocument();
    expect(logo).toHaveAttribute('src', 'mock-logo');

    const projectsIcon = screen.getByAltText('Projects');
    expect(projectsIcon).toBeInTheDocument();
    expect(projectsIcon).toHaveAttribute('src', 'mock-projects-icon');

    const projectLink = screen.getByRole('link', { name: /projects/i });
    expect(projectLink).toHaveAttribute('href', ROUTE.PROJECT);

    const settingIcon = screen.getByAltText('Setting');
    expect(settingIcon).toBeInTheDocument();
    expect(settingIcon).toHaveAttribute('src', 'mock-setting-icon');

    const settingLink = screen.getByRole('link', { name: /setting/i });
    expect(settingLink).toHaveAttribute('href', ROUTE.SETTING);
  });
});
