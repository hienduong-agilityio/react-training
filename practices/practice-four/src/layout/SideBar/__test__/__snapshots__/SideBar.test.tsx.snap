// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex bg-primary-700"
  >
    <div
      class="w-[74px] bg-primary-700 h-screen p-4 relative duration-300"
    >
      <div
        class="flex items-center"
      >
        <a
          href="/projects"
        >
          <img
            alt="Logo"
            class="cursor-pointer duration-500"
            src="mock-logo"
          />
        </a>
      </div>
      <ul
        class="pt-10 flex flex-col items-center justify-center gap-5"
      >
        <li
          class="rounded-md items-center mt-2 focus:bg-primary-900 cursor-pointer hover:border-primary-100 hover:bg-primary-600"
        >
          <a
            href="/projects"
          >
            <img
              alt="Projects"
              class="p-2"
              src="mock-projects-icon"
            />
          </a>
        </li>
        <li
          class="rounded-md items-center mt-2 focus:bg-primary-900 cursor-pointer hover:border-primary-100 hover:bg-primary-600"
        >
          <a
            href="/settings"
          >
            <img
              alt="Setting"
              class="p-2"
              src="mock-setting-icon"
            />
          </a>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
